### PR TITLE
[Gluten-core] Get adaptive execution context known through checking thread stack trace

### PR DESCRIPTION
**Pull request by PHILO-HE**
## What changes were proposed in this pull request?

Currently, there are some differences in applying columnar rule, depending on whether adaptive execution is taking effect. So it is important to check whether the current execution goes into AQE context. Previously, we achieved the check by porting a piece of spark's source code, which is not a good approach. As that code logic needs our careful maintenance, especially in supporting multiple versions of spark. On the other hand, it breaks the principle of encapsulation, as Hongze pointed out.
Here, we are proposing to get AQE context known through checking thread stack trace, assuming the stack trace is stable. This approach may be tricky, but is more reliable and easier to maintain than the current approach.

### CODE REVIEW

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala [hunk 1]
```diff
@@ -471,25 +473,36 @@ case class ColumnarOverrideRules(session: SparkSession)
...
   override def preColumnarTransitions: Rule[SparkPlan] = plan => PhysicalPlanSelector.
     maybe(session, plan) {
       var overridden: SparkPlan = plan
       val startTime = System.nanoTime()
-      supportAdaptive = AdaptiveSparkPlanUtil.supportAdaptiveWithExchangeConsidered(plan)
+      this.isLeafPlanExchange = plan match {
+        case _: Exchange => true
+        case _ => false
+      }
+      val traces = Thread.currentThread.getStackTrace()
```
Comment (Yohahaha) [inline gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:494]: Java call parentheses in scala can be removed.

Comment (PHILO-HE) [reply gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala]: Thanks for your comment! Just fixed in the latest commit.

**Comment by PHILO-HE** [comment]
@zhztheplayer, @marin-ma, please help review this PR. It has the same purpose for improving our AQE check logic as a previous PR.

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala [hunk 2]
```diff
@@ -471,25 +473,37 @@ case class ColumnarOverrideRules(session: SparkSession)
...
   override def preColumnarTransitions: Rule[SparkPlan] = plan => PhysicalPlanSelector.
     maybe(session, plan) {
       var overridden: SparkPlan = plan
       val startTime = System.nanoTime()
-      supportAdaptive = AdaptiveSparkPlanUtil.supportAdaptiveWithExchangeConsidered(plan)
+      this.isLeafPlanExchange = plan match {
+        case _: Exchange => true
+        case _ => false
+      }
+      val traceElements = Thread.currentThread.getStackTrace
```
Comment (marin-ma) [inline gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:494]: If we have to do this in a hacky way, it's probably better to move the depth counting part into shim layer.

Comment (PHILO-HE) [reply gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:494]: Thanks for your comment!
I guess this calling stack is stable. And at least for our currently supported spark versions, there is no difference. I think if in the future other spark versions change it, we can differentiate the impl. in shim layers. Right?

Comment (marin-ma) [reply gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:494]: Sure. To make the code more readable, maybe we can name the number "13" to something like "AqeStatcktraceDepth"?

Comment (PHILO-HE) [reply gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:496]: Yes, just made some changes in the latest commit. Thanks!

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala [hunk 3]
```diff
@@ -471,25 +475,37 @@ case class ColumnarOverrideRules(session: SparkSession)
...
+      assert(traceElements.length > 13, "More than 13 stack trace elements are expected!")
```
Comment (marin-ma) [inline gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:497]: Please also replace with the constant name here.

**Comment by marin-ma** [comment]
LGTM.

**Comment by PHILO-HE** [comment]
@marin-ma, please approve if you have no other suggestion. Thanks! And if no more comment from others, will merge the patch soon.

**Review by marin-ma — APPROVED** [review]

**Pull request merged**